In [1]:
import numpy as np
import keras

Using TensorFlow backend.


In [2]:
from keras.datasets import imdb

In [67]:
import nltk

In [68]:
from nltk.corpus import stopwordspwords

In [9]:
from keras.models import load_model

In [3]:
from keras.models import Sequential
from keras.layers import Dense, LSTM, Dropout
from keras.layers.embeddings import Embedding

In [4]:
import h5py

In [5]:
from keras.preprocessing import sequence

In [4]:
imdb?

Keras already has the iMdb dataset Built-in.The dataset 25000 training and testing reviews each. Also the data is formatted with each word replacing the rank of thier occurance 

In [5]:
imdb.load_data()

16949248/17464789 [============================>.] - ETA: 0s

((array([ list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 22665, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 21631, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 19193, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 10311, 8, 4, 107, 117, 5952, 15, 256, 4, 31050, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 12118, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
         list([1, 1

get_word_index contains the dictionary of all the words with their occurance rank

In [12]:
word_indices = imdb.get_word_index()

In [6]:
word_indices

{'madness': 2998,
 'sacarstic': 78004,
 'prejudice': 5518,
 "characteristic's": 60890,
 'tolmekians': 82327,
 'glands': 32509,
 'tink': 38154,
 'creeds': 28028,
 'airspeed': 81294,
 'naziism': 55785,
 'bauman': 41362,
 'script': 226,
 'clever': 1093,
 'hehehe': 69670,
 "tower's": 60459,
 'obama': 35587,
 'overcame': 20295,
 'mainstay': 19099,
 'mohave': 54686,
 'jez': 61542,
 'unfolds': 4183,
 'carnosaur': 14328,
 'brunt': 20716,
 "hiraizumi's": 71767,
 'naples': 56221,
 'screwier': 63150,
 "dung's": 83024,
 'stabs': 8525,
 'kitbag': 56592,
 'unsettling': 4330,
 'affirmations': 41328,
 'norrland': 44484,
 "'proper": 69694,
 'maudette': 70834,
 'oaters': 28134,
 'cacoyanis': 68252,
 'groundwork': 29908,
 'lashes': 20972,
 'pinkie': 45922,
 'bobs': 29324,
 'pleb': 65895,
 "1840'": 80678,
 'foxed': 56444,
 'rosenberg': 28563,
 "agee's": 63050,
 "'radio": 42321,
 'oz': 4490,
 'ester': 35105,
 'antionioni': 82312,
 'renyolds': 40817,
 'corenblith': 73401,
 "'gods'": 71871,
 'firecrackers': 

In [16]:
INDEX_FROM=3   # word index offset
train,test = imdb.load_data(index_from=INDEX_FROM)
x_train, y_train =  train
x_test,y_test = test

In [18]:
x_train.shape,x_test.shape

((25000,), (25000,))

In [19]:
word_to_id = {k:(v+INDEX_FROM) for k,v in word_indices.items()}
word_to_id["SPACE"] = 0
word_to_id["<START>"] = 1
word_to_id["END"] = 2
 

In [22]:
id_to_word = {value:key for key,value in word_to_id.items()}
print(" ".join(id_to_word[1] for i in x_train[1]))

<START> big hair big boobs bad music and a giant safety pin these are the words to best describe this terrible movie i love cheesy horror movies and i've seen hundreds but this had got to be on of the worst ever made the plot is paper thin and ridiculous the acting is an abomination the script is completely laughable the best is the end showdown with the cop and how he worked out who the killer is it's just so damn terribly written the clothes are sickening and funny in equal measures the hair is big lots of boobs bounce men wear those cut tee shirts that show off their stomachs sickening that men actually wore them and the music is just synthesiser trash that plays over and over again in almost every scene there is trashy music boobs and paramedics taking away bodies and the gym still doesn't close for bereavement all joking aside this is a truly bad film whose only charm is to look back on the disaster that was the 80's and have a good old laugh at how bad everything was back then


In [6]:
top_words = 5000
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=top_words)

# First model with One LSTM Layer without dropout

In [9]:
model = Sequential()
model.add(Embedding(top_words,64))
model.add(LSTM(100))
model.add(Dense(1, activation="sigmoid"))
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 64)          320000    
_________________________________________________________________
lstm_1 (LSTM)                (None, 100)               66000     
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 101       
Total params: 386,101
Trainable params: 386,101
Non-trainable params: 0
_________________________________________________________________


In [8]:
x_train = sequence.pad_sequences(X_train, maxlen=500)
x_test = sequence.pad_sequences(X_test, maxlen=500)

In [17]:
x_train.shape

(25000, 500)

In [10]:
model.fit(x_train, y_train, epochs=3, batch_size=128)

Epoch 1/3
25000/25000 [==============================] - 154s - loss: 0.4529 - acc: 0.7797   
Epoch 2/3
25000/25000 [==============================] - 136s - loss: 0.2780 - acc: 0.8895   
Epoch 3/3
25000/25000 [==============================] - 134s - loss: 0.2515 - acc: 0.9016   


In [ ]:
scores = model.evaluate(x_test, y_test, verbose=0)
print("Accuracy: %.2f%%" % (scores[1]*100))

Accuracy: 86.88%


As we can see The validation accracy is 86.88 but the model is prone to overfitting hence we use droput layers

# Another model with 2 LSTM layers and also dropout layer

In [23]:
model2 = Sequential()
model2.add(Embedding(top_words,64))
model2.add(LSTM(128, return_sequences=True, recurrent_dropout=0.1, dropout=0.2))
# model2.add(Dropout(0.2))
model2.add(LSTM(64, recurrent_dropout=0.1, dropout=0.2))
# model2.add(Dropout(0.2))
model2.add(Dense(1,activation="sigmoid"))
model2.compile(loss="binary_crossentropy",optimizer="adam",metrics=["accuracy"])
model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_9 (Embedding)      (None, None, 64)          320000    
_________________________________________________________________
lstm_10 (LSTM)               (None, None, 128)         98816     
_________________________________________________________________
lstm_11 (LSTM)               (None, 64)                49408     
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 65        
Total params: 468,289
Trainable params: 468,289
Non-trainable params: 0
_________________________________________________________________


In [25]:
model2.fit(x_train, y_train, epochs=5, batch_size=128)

Epoch 1/5
25000/25000 [==============================] - 269s - loss: 0.4783 - acc: 0.7621   
Epoch 2/5
25000/25000 [==============================] - 266s - loss: 0.4495 - acc: 0.7984   
Epoch 3/5
25000/25000 [==============================] - 266s - loss: 0.3701 - acc: 0.8490   
Epoch 4/5
25000/25000 [==============================] - 263s - loss: 0.2853 - acc: 0.8882   
Epoch 5/5
25000/25000 [==============================] - 266s - loss: 0.2938 - acc: 0.8832   


Saving the model Using h5py

In [27]:
model2.save("model2.h5")
model.save("model.h5")

In [10]:
model1 = load_model("model.h5")
model2 = load_model("model2.h5")

In [ ]:
scores = model2.evaluate(x_test, y_test, verbose=0)
print("Accuracy: %.2f%%" % (scores[1]*100))

Accuracy: 85.47%


The validation accuracy might be less but this model is less prone to overfitting and is a more generalized model